spec/unit/pops/validator/validator_spec.rb

```diff
@@ -229,6 +229,20 @@ def validate(model)
     end
   end
 
+  context 'for parameter defaults' do
+    ['class', 'define', 'function'].each do |word|
+      it "should not permit assignments in #{word} parameter default expressions" do
+        expect { parse("#{word} foo($a = $x = 10) {}") }.to raise_error(Puppet::ParseErrorWithIssue, /Syntax error at '='/)
+      end
+    end
+
+    ['class', 'define', 'function'].each do |word|
+      it "should not permit assignments in #{word} parameter default nested expressions" do
+        expect(validate(parse("#{word} foo($a = [$x = 10]) {}"))).to have_issue(Puppet::Pops::Issues::ILLEGAL_ASSIGNMENT_CONTEXT)
+      end
+    end
+  end
+
```

Comment:
what if it is an assignment to a parameter, and to itself. $a = $a, and $a = 10, $b = ($a = 10). Also, if parameter default expression contains a lambda, assignments in the lambdas local scope should be allowed (so, test is needed for that).